File: LongLib/src/main/java/com/hloong/lib/longlog/LongStackTraceUtil.java
package com.hloong.lib.longlog;

public class LongStackTraceUtil {
    /**
     * To get clip stack information
     * @param stackTrace
     * @param ignorePackage
     * @param maxDepth
     * @return
     */
    public static StackTraceElement[] getCroppedRealStackTrack(StackTraceElement[] stackTrace,String ignorePackage,int maxDepth){
        return cropStackTrace(getRealStackTrack(stackTrace,ignorePackage),maxDepth);
    }

    /**
     * To get the stack information except ignore the package name
     * @param stackTrace
     * @param ignorePackage
     * @return
     */
    private static StackTraceElement[] getRealStackTrack(StackTraceElement[] stackTrace,String ignorePackage){
        int ignoreDepth = 0;
        int allDepth = stackTrace.length;
        String className;
        for (int i = allDepth-1; i >=0 ; i--) {
            className = stackTrace[i].getClassName();
            if (ignorePackage != null && className.startsWith(ignorePackage)){
                ignoreDepth = i + 1;
                break;
            }
        }
        int realDepth = allDepth - ignoreDepth;
        StackTraceElement[] realStack = new StackTraceElement[realDepth];
        System.arraycopy(stackTrace,ignoreDepth,realStack,0,realDepth);
        return realStack;
    }

    /**
     * Clipping stack information
     * @param callStack
     * @param maxDepth
     * @return
     */
    private static StackTraceElement[] cropStackTrace(StackTraceElement[] callStack,int maxDepth){
       int realDepth = callStack.length;
       if (maxDepth >0){
           realDepth = Math.min(maxDepth,realDepth);
       }
       StackTraceElement[] realStack = new StackTraceElement[realDepth];
       System.arraycopy(callStack,0,realStack,0,realDepth);
       return realStack;
    }

}

File: LongLib/src/main/java/com/hloong/lib/longlog/base/LongStackTraceUtil.kt
package com.hloong.lib.longlog.base

import kotlin.math.min

object LongStackTraceUtil {
    /**
     * To get clip stack information
     * @param stackTrace
     * @param ignorePackage
     * @param maxDepth
     * @return
     */
    @JvmStatic
    fun getCroppedRealStackTrack(
        stackTrace: Array<StackTraceElement>,
        ignorePackage: String?,
        maxDepth: Int
    ): Array<StackTraceElement?> {
        return cropStackTrace(getRealStackTrack(stackTrace, ignorePackage), maxDepth)
    }

    /**
     * To get the stack information except ignore the package name
     * @param stackTrace
     * @param ignorePackage
     * @return
     */
    private fun getRealStackTrack(
        stackTrace: Array<StackTraceElement>,
        ignorePackage: String?
    ): Array<StackTraceElement?> {
        var ignoreDepth = 0
        val allDepth = stackTrace.size
        var className: String
        for (i in allDepth - 1 downTo 0) {
            className = stackTrace[i].className
            if (ignorePackage != null && className.startsWith(ignorePackage)) {
                ignoreDepth = i + 1
                break
            }
        }
        val realDepth = allDepth - ignoreDepth
        val realStack = arrayOfNulls<StackTraceElement>(realDepth)
        System.arraycopy(stackTrace, ignoreDepth, realStack, 0, realDepth)
        return realStack
    }

    /**
     * Clipping stack information
     * @param callStack
     * @param maxDepth
     * @return
     */
    private fun cropStackTrace(
        callStack: Array<StackTraceElement?>,
        maxDepth: Int
    ): Array<StackTraceElement?> {
        var realDepth = callStack.size
        if (maxDepth > 0) {
            realDepth = min(maxDepth, realDepth)
        }
        val realStack = arrayOfNulls<StackTraceElement>(realDepth)
        System.arraycopy(callStack, 0, realStack, 0, realDepth)
        return realStack
    }
}